### PR TITLE
Windows connection fixes

### PR DESCRIPTION
**Pull request by dlech**
This fixes some issues with the Windows disconnect callback implementation (e.g. #312 and #313). 

It also contains further improvements for managing the connection lifecycle.

### CODE REVIEW

File: bleak/backends/dotnet/client.py [hunk 1]
```diff
@@ -168,79 +170,117 @@ async def connect(self, **kwargs) -> bool:
...
 
+        # Called on disconnect event or on failure to connect.
+        def handle_disconnect():
+            if self._connection_status_changed_token:
+                self._requester.remove_ConnectionStatusChanged(
+                    self._connection_status_changed_token
+                )
+                self._connection_status_changed_token = None
+
+            if self._requester:
```
Comment (hbldh) [inline bleak/backends/dotnet/client.py:181]: Here we (potentially) dispose of the requester before the session, which is opposite from how it is done in `disconnect`. Will that be a problem do you think?

**Review by hbldh — APPROVED** [review]

**Comment by hbldh** [comment]
This is so good. I will add a commit changing event handling in ``discovery.py`` and ``scanner.py`` as well.

I was looking at `GattSession` a while back and didn't see the value of adding it then. I see it now. It was definitely a missing piece in the implementation.

Please add the changes to the `CHANGELOG` as well!

File: bleak/backends/dotnet/client.py [hunk 2]
```diff
@@ -168,79 +170,117 @@ async def connect(self, **kwargs) -> bool:
...
-            ):
-                loop.call_soon_threadsafe(self._disconnected_callback, self)
-
-        self._requester.ConnectionStatusChanged += _ConnectionStatusChanged_Handler
+            loop.call_soon_threadsafe(
+                handle_connection_status_changed, sender.ConnectionStatus
+            )
 
-        # Obtain services, which also leads to connection being established.
-        services = await self.get_services()
```
Comment (hbldh) [inline bleak/backends/dotnet/client.py:186]: If `await self.get_services()` is not called in `connect`, it is required of the user to call it before trying to do any other operations. It should be done here to make sure that the `BleakClient` is ready for anything after the connection.

**Review by hbldh — CHANGES_REQUESTED** [review]

File: bleak/backends/dotnet/client.py (hunk 2 above)
Comment (dlech) [reply bleak/backends/dotnet/client.py:186]: Fixed. Thanks.

**Review by hbldh — APPROVED** [review]

**Pull request merged**